### PR TITLE
Add files via upload

### PR DESCRIPTION
**Pull request by Natakass**

### CODE REVIEW

File: homework_(2) (1).ipynb [hunk 1]
```diff
@@ -0,0 +1,1918 @@
+{
```
Comment (Pretzandalf) [inline homework_(2) (1).ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #13.        attention = F.softmax(torch.matmul(query, key.transpose(-2, -1)) / (query.size(-1) ** 0.5), dim=-1 )
</pre><p>От себя на будущие добавлю, старайся использовать nn вместо F. Функции в nn содержат под капотом те же F функции, но они сделаны так чтобы удобнее работать с обучаемыми параметрами и наример с батчами. Скажем, так что nn более высокоуровневый. (здесь вообще не важно nn или F, просто как фанфакт))</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Natakass/transormers/pull/1/discussion/#comment-2190984168'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='12'/>

Comment (Pretzandalf) [inline homework_(2) (1).ipynb:1]: <p>Моя шиза говорила мне что так сделать правильнее,</p>

<pre class="ql-syntax" spellcheck="false">    outputs = self.dropout1(self.norm1(x + outputs))

&nbsp;&nbsp;&nbsp;&nbsp;outputs = self.norm2(outputs + self.feedforward(outputs))
</pre><p>но по факту твое решение правильнее.</p><p>-----------------------------------------------------------------------------------------------------------------------------------------</p>

<pre class="ql-syntax" spellcheck="false">norm(x + dropout(sublayer(x))
</pre><p>в трансформерах используется такая последовательность, в том же коде берта:</p>

<pre class="ql-syntax" spellcheck="false">&nbsp;&nbsp;def forward(self, hidden_states: torch.Tensor, input_tensor: torch.Tensor) -&gt; torch.Tensor:
&nbsp;&nbsp;&nbsp;&nbsp;hidden_states = self.dense(hidden_states)
&nbsp;&nbsp;&nbsp;&nbsp;hidden_states = self.dropout(hidden_states)
&nbsp;&nbsp;&nbsp;&nbsp;hidden_states = self.LayerNorm(hidden_states + input_tensor)
&nbsp;&nbsp;&nbsp;&nbsp;return hidden_states
</pre><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Natakass/transormers/pull/1/discussion/#comment-2190984174'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='29'/>

Comment (Pretzandalf) [inline homework_(2) (1).ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #21.            self.encoder = nn.ModuleList(encoder_blocks) # TODO
</pre><p>можно было использовать nn.Sequential, но твой подход тоже верный. (Модульлист обычно используют в сложных ситуациях, когда разные параметры в блоках, не все надо вызывать и тд.</p><p> </p><p> </p><p>------------------------------------------------------------------------------------------------------------</p>

<pre class="ql-syntax" spellcheck="false">    self.cls_token = nn.Parameter(torch.randn(embed_dim)) # TODO create vector of size (embed_dim,) from N(0, 1)

&nbsp;&nbsp;&nbsp;&nbsp;cls = self.cls_token.expand(x.size(0), 1, x.size(2))
&nbsp;&nbsp;&nbsp;&nbsp;x = torch.cat([cls.to(device), x], dim=1)
&nbsp;&nbsp;&nbsp;&nbsp;x = self.positional_encoding(x)
</pre><p>радуют эти две строчки, все прям как в берте сделано. nn.Parameter на клс и pe после добавления cls</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Natakass/transormers/pull/1/discussion/#comment-2190984175'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='51'/>

Comment (Pretzandalf) [inline homework_(2) (1).ipynb:1]: <p>С Linformer все отлично.</p><p>Очень порадовала твоя работа. Классно, что модельки добивают до акк ~1.0</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Natakass/transormers/pull/1/discussion/#comment-2190984178'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='92'/>

Comment (dpaleyev) [reply homework_(2) (1).ipynb:1]: <p>Я не согласен, Наташа правильно использовала ModuleList. Здесь нельзя использовать Sequential, так как нам надо сохранять выходы attention maps которые не пробрасываются в следующие слои</p>
